vhdl_ls/src/vhdl_server.rs

```diff
@@ -398,6 +404,23 @@ impl<T: RpcChannel + Clone> InitializedVHDLServer<T> {
         self.text_document_declaration(params)
     }
 
+    pub fn text_document_hover(&mut self, params: &TextDocumentPositionParams) -> Option<Hover> {
+        self.project
+            .get_source(&uri_to_file_name(&params.text_document.uri))
+            .and_then(|source| {
+                self.project
+                    .search_reference(&source, from_lsp_pos(params.position))
```

Comment:
Isn't the search done twice here? Why does `format_declaration` need to search again when we already found it with `search_reference`.  I assume it was because `search_reference` did not return the AST-value but in that case this function should be generalized to avoid searching twice.

Reply:
There are two searches done here, yes, but they are different. The first searches for the token at the position provided over the language server protocol, and retrieves the referenced declaration from it. At the moment, that only provides the position of the declaration, so we need to search for the declaration itself in order to retrieve the AST node that we can actually format.
This second search could possibly be rendered obsolete if Reference were instead the ID of a NamedEntity (as you suggest below), but that would require two things:
1. Lookup of the named entity by id. This may already exist, in fact I assume it must, but I haven't found it yet.
2. Enough information within the named entity in order to reconstruct the VHDL code. This would probably be most easily achieved by referencing back into the AST in order to avoid duplication of information.
Unfortunately, that second point is beyond my understanding of the code atm.